go-app-ussd_sassa_registration.js

```diff
@@ -0,0 +1,203 @@
+            dob;
+          var validLuhn = function (content) {
+            return (
+              content
+                .split("")
+                .reverse()
+                .reduce(function (sum, digit, i) {
+                  return (
+                    sum +
+                    _.parseInt(
+                      i % 2 ? [0, 2, 4, 6, 8, 1, 3, 5, 7, 9][digit] : digit
+                    )
+                  );
+                }, 0) %
+                10 ==
+              0
+            );
+          };
+          if (
+            !match ||
+            !validLuhn(content) ||
+            !(dob = new moment(match[1], "YYMMDD")) ||
+            !dob.isValid() ||
+            !dob.isBetween(
+              today.clone().add(-130, "years"),
+              today.clone().add(-17, "years")
+            )
+          ) {
+            return error_msg;
+          } else {
+            return new JsonApi(self.im)
+              .post(self.im.config.sassa_api.url + "/api/v1/check_id_number", {
+                data: {
+                  id_number: content,
+                },
+                headers: {
+                  Authorization: ["Token " + self.im.config.sassa_api.token],
+                  "User-Agent": ["Jsbox/SASSA-Registration-USSD"],
+                },
+              })
+              .then(
+                function (response) {
+                  if (!response.data.valid || response.data.underage) {
+                    return error_msg;
+                  } else if (response.data.existing) {
+                    return $(
+                      [
+                        "An application for {{ id_number }} has already been made. The status is: {{ status }}.",
+                        "To appeal, call 0800 002 9999",
+                      ].join("\n")
+                    ).context({
+                      id_number: content,
+                      status: response.data.status,
+                    });
+                  }
+                },
+                function (e) {
+                  // Go to error state after 3 failed HTTP requests
+                  opts.http_error_count =
+                    _.get(opts, "http_error_count", 0) + 1;
+                  if (opts.http_error_count === 3) {
+                    self.im.log.error(e.message);
+                    return self.states.create("__error__", {
+                      return_state: name,
+                    });
+                  }
+                  return self.states.create(name, opts);
+                }
+              );
+          }
+        },
+        next: "state_grant",
+      });
+    });
+
+    self.states.creators.__error__ = function (name, opts) {
+      var return_state = opts.return_state || "state_start";
+      return new EndState(name, {
+        next: return_state,
+        text: $(
+          "Sorry, something went wrong. We have been notified. Please try again later"
```

Comment:
Is there a cooling off period that we can implement? This would be in the case where someone tries more than 3 times.

Reply:
I'll make a note to try and implement at a later stage.